### PR TITLE
Ensure user data is fully deleted on Multisite installs

### LINKED ISSUE #715: User signups should be marked `active=0` after account is deleted
Issue raised via https://core.trac.wordpress.org/ticket/43232

I have tested this and can reproduce the issue in the `develop` branch:

1. Create a new user at `wp-signup.php`
2. In admin, delete this new user
3. Try to create the user again with the sane username and email address - see the image below
4. Note that an entry remains in the `cp_signups` table

Screenshot 2021-04-25 at 16 51 10

Switch to this branch and repeat the above steps.
Note this time that after the user is deleted the entry in the `cp_signups` table also deleted.
Note also that the same user details can immediately be used to sign up again.

Screenshot 2021-04-25 at 16 58 39

_Originally posted by @mattyrob in https://github.com/ClassicPress/ClassicPress/issues/593#issuecomment-826347354_

### PR DESCRIPTION
**Pull request by mattyrob**
## Description
Backpost of upstream bug fix that closes #715 

## Motivation and context
Multisite: Delete the signup entry when a user is deleted from the database.
>     
>     When a user has registered with incorrect data and got the account deleted immediately, it currently wasn't possible to register with the same login or email address again due to the existing signup entry. They had to wait for two days until the entry gets automatically deleted. Now the associated signup entry gets deleted as part of the account deletion.
>     
>     Fixes #43232.

Unti test back post not applying cleanly - some revisions need looking at manually.

## How has this been tested?
Unit tests are included in patch

## Types of changes
- Bug fix

### CODE REVIEW

**Comment by bahiirwa** [comment]
Travis is breaking because of the merge conflicts. @mattyrob are you in position to resolve those?

**Comment by mattyrob** [comment]
I've just pushed an update to address the merge conflicts.

**Comment by bahiirwa** [comment]
Fixes look good.

**Comment by mattyrob** [comment]
I have tested this and can reproduce the issue in the `develop` branch:

1. Create a new user at `wp-signup.php`
2. In admin, delete this new user
3. Try to create the user again with the sane username and email address - see the image below
4. Note that an entry remains in the `cp_signups` table

Screenshot 2021-04-25 at 16 51 10

Switch to this branch and repeat the above steps.
Note this time that after the user is deleted the entry in the `cp_signups` table also deleted.
Note also that the same user details can immediately be used to sign up again.

Screenshot 2021-04-25 at 16 58 39

**Comment by bahiirwa** [comment]
@mattyrob I pushed an issue #715 and referenced your test. I was puzzled at the change.

**Comment by mattyrob** [comment]
> I was puzzled at the PR. @mattyrob I pushed an issue and referenced your test.
> 
> Can you add a PR closes #715 in description of the PR?

I don't seem to have the right privileges for this but I've tagged the issue with a 'Has PR' label.

**Comment by nylen** [comment]
You can edit the original description to add "closes #715" or "fixes #715" in the text, that will do it.

**Comment by nylen** [comment]
I noticed that the unit test didn't merge over properly. This is probably because just backporting the WP changeset 48315 added a huge number of merge conflicts. This was happening because [WP's massive reformatting commit](https://github.com/ClassicPress/ClassicPress/commit/8f95800d52c1736d651ae6e259f90ad4a0db2c3f) re-indented this whole file, and `git` doesn't know how to merge changes involving the same code but different indentation levels or other formatting.

This is probably not the last time we'll see this issue, so the best way I could think to fix this was to selectively apply the changes in that commit to that one file:

```
git show 8f95800d5 tests/phpunit/tests/multisite/wpmuValidateUserSignup.php | git apply
git add tests/phpunit/tests/multisite/wpmuValidateUserSignup.php
git commit
# Write a commit message that indicates why this specific change was made and where it came from
```

If the patch doesn't apply cleanly then that `git apply` command will just fail and you have to use something that will trigger git's conflict resolution mechanism instead. This uses more powerful `git` commands (actually the same ones that the git and Linux kernel maintainers use to send and apply patches via email) and is a bit more complicated:

```
git format-patch --stdout -1 8f95800d5 tests/phpunit/tests/multisite/wpmuValidateUserSignup.php | git am
git commit --amend
# Edit the commit message to something that better indicates why this specific change was made and where it came from
```

Either way the result should be a commit that looks something like commit fc044448a in this PR. After that, the backport (using the `-c` flag to the backport script to stay on the same branch) should go through cleanly.

**Comment by nylen** [comment]
Also the `@ticket` annotation is a [silly non-standard WP thing that we removed](https://github.com/ClassicPress/ClassicPress/pull/43), so whenever that appears in a backported changeset, it should be changed as follows:

```diff
- @ticket 12345
+ @see https://core.trac.wordpress.org/ticket/12345
```

File: src/wp-includes/ms-functions.php [hunk 1]
```diff
@@ -1118,6 +1118,21 @@ function wpmu_activate_signup($key) {
 	return array('blog_id' => $blog_id, 'user_id' => $user_id, 'password' => $password, 'title' => $signup->title, 'meta' => $meta);
 }
 
+/**
+ * Deletes am associated signup entry when a user is deleted from the database.
```
Comment (nylen) [inline src/wp-includes/ms-functions.php:1122]: This typo can stay for now, and if WP ever fixes it then we can backport the fix. This should be an easy way to get contributor credits over there also.

**Review by nylen — APPROVED** [review]
Thanks for the PR, looks good.

File: src/wp-includes/ms-functions.php (hunk 1 above)
Comment (bahiirwa) [reply src/wp-includes/ms-functions.php:1122]: I have reported this back upstream https://core.trac.wordpress.org/ticket/52628#comment:24

**Pull request merged**